hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java

```diff
@@ -94,9 +103,86 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 		}
 	}
 
+	private <T extends AST> List<T> findAllNodes(AST node, Class<T> clazz) {
+		ArrayList<T> found = new ArrayList<>();
+		doFindAllNodes( node, clazz, found );
+		return found;
+	}
+
+	private <T extends AST> void doFindAllNodes(AST node, Class<T> clazz, List<T> found) {
+		if ( clazz.isAssignableFrom( node.getClass() ) ) {
+			found.add( (T) node );
+		}
+		if ( node.getFirstChild() != null ) {
+			doFindAllNodes( node.getFirstChild(), clazz, found );
+		}
+		if ( node.getNextSibling() != null ) {
+			doFindAllNodes( node.getNextSibling(), clazz, found );
```

Comment:
@beikov This is the code that caused the `StackOverflowError` for @heruan . The code seems to be fine under the assumption of a tree structure. I'm not sure in what scenario the AST could possibly lose this tree property... Does this ring a bell for you? Perhaps the query was just fairly complex and the stack relatively limited for a recursive implementation to work on the number of nodes in the tree. Perhaps then we should rewrite the implementation to do a queue based BFS instead.


```
java.lang.StackOverflowError: null
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:113) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
```